### ISSUE TITLE
torch.sum(x,x,2) not returning correct value

### ISSUE THREAD

**Issue by ghostcow** [torch/torch7#830]
When giving torch.sum the same tensor as the one summed, result has correct shape but incorrect values.

Reproduce:

```
x=torch.range(1,9):reshape(3,3)
print(x)
print(torch.sum(x,2))
print(torch.sum(x,x,2))
```

Output:

```
 1  2  3
 4  5  6
 7  8  9
[torch.DoubleTensor of size 3x3]

  6
 15
 24
[torch.DoubleTensor of size 3x1]

 1
 2
 3
[torch.DoubleTensor of size 3x1]
```

My use case is to simply reduce allocations

**Comment by soumith** [comment]
i dont think we can even expect this to be correct. The best thing we can do is give an error if result tensor = src tensor

**Comment by ghostcow** [comment]
yeah i guess one would need a buffer anyway to store results thus defeating the purpose :(
where would it be efficient to place this assertion?

**Comment by soumith** [comment]
probably here is best: https://github.com/torch/torch7/blob/master/lib/TH/generic/THTensorMath.c#L1139
